pkg/channel/jetstream/dispatcher/consumer.go

```diff
@@ -19,122 +19,226 @@ package dispatcher
+	consumerInfo, err := consumer.Info(ctx)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get consumer info: %w", err)
 	}
 
-	c.closed = true
+	logger := logging.FromContext(ctx)
+
+	return &Consumer{
+		dispatcher:       dispatcher,
+		reporter:         reporter,
+		channelNamespace: channelNamespace,
+
+		natsConsumer:     consumer,
+		natsConsumerInfo: consumerInfo,
+
+		logger: logger,
+
+		sub: subscription,
 
-	// TODO: should we wait for messages to finish and c.jsSub.IsValid() to return false?
-	return c.jsSub.Drain()
+		closing: make(chan struct{}),
+		closed:  make(chan struct{}),
+	}, nil
 }
 
-func (c *Consumer) MsgHandler(msg *nats.Msg) {
-	logger := c.logger.With(zap.String("msg_id", msg.Header.Get(nats.MsgIdHdr)))
-	ctx := logging.WithLogger(c.ctx, logger)
-	tickerCtx, tickerCancel := context.WithCancel(c.ctx)
+// Start begins the consumer and handles messages until Close is called. This method is blocking and
+// will return an error if the consumer fails prematurely. A nil error will be returned upon being
+// stopped by the Close method.
+func (c *Consumer) Start() error {
+	if err := c.checkStart(); err != nil {
+		return err
+	}
 
-	tickerDone := make(chan struct{})
+	defer close(c.closed)
 
-	go func() {
-		defer close(tickerDone)
+	ctx := logging.WithLogger(context.Background(), c.logger)
 
-		// TODO(dan-j): this should be a fraction of the Consumer's AckWait
-		ticker := time.NewTicker(10 * time.Second)
-		defer ticker.Stop()
+	var wg sync.WaitGroup
 
-		for {
-			select {
-			case <-tickerCtx.Done():
-				return
-			case <-ticker.C:
-				if err := msg.InProgress(nats.Context(tickerCtx)); err != nil && !errors.Is(err, context.Canceled) {
-					logging.FromContext(ctx).Errorw("failed to mark message as in progress", zap.Error(err))
-				}
+	// wait for handlers to finish
+	defer wg.Wait()
+
+	// all messages are attached with this ctx, upon closing the consumer, we will cancel this ctx
+	// so that all pending messages are cancelled. This should cause any pending requests be
+	// cancelled (which also results in a nack) and the batch to be drained.
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	for {
+		batch, err := c.natsConsumer.FetchNoWait(FetchBatchSize)
+		if err != nil {
+			return err
+		}
+
+		if err := c.consumeMessages(ctx, batch, &wg); err != nil {
+			if errors.Is(err, io.EOF) {
+				return nil
 			}
+
+			return err
 		}
-	}()
+	}
+}
 
-	go func() {
-		var result protocol.Result
+// consumeMessages consumes messages from the batch and forwards them to the queue. The enqueued
+// message includes a context which will cancel after the AckWait duration of the consumer.
+//
+// This method returns once the MessageBatch has been consumed, or upon a call to Consumer.Close.
+// Returning as a result of Consumer.Close results in an io.EOF error.
+func (c *Consumer) consumeMessages(ctx context.Context, batch jetstream.MessageBatch, wg *sync.WaitGroup) error {
+	for {
+		select {
```

Comment:
This allows us to stop immediately when someone calls `Consumer.Close()`, but yeah you could continue to process the batch since `ctx` will be cancelled anyway, just if you've got a batch of 100 when closing, you might end up with 100 context deadline exceeded errors.